### ISSUE TITLE
No covariance in graph ~> operator

### ISSUE THREAD

**Issue by 2m** [akka/akka#16997]
This does not compile
``` scala
FlowGraph.closed() { implicit b =>
  val merge = b.add(Merge[AnyRef](inputPorts = 2))
  Source.single("") ~> merge // compile error
  Source.single('a) ~> merge // compile error
  merge ~> Sink.head[AnyRef]()
}
```

while this does
``` scala
FlowGraph.closed() { implicit b =>
  val merge = b.add(Merge[AnyRef](inputPorts = 2))
  Source.single("") ~> merge.in(0)
  Source.single('a) ~> merge.in(1)
  merge ~> Sink.head[AnyRef]()
}
```

It seems that covariance is missing on the ~> type parameters in  https://github.com/akka/akka/blob/ab5702e65ce1e78a88bc448b4ca2798464ce84d3/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala#L289

**Comment by rkuhn** [comment]
No, in fact it is only missing on the boilerplate-generated Shape classes (UniformFanInShape in this case).

**Comment by 2m** [comment]
This problem from the OP has been solved in #17636. But there are still problems with ~> and variance as indicated [here](https://github.com/2m/akka/commit/2e5beb2d7d7d9eedaf83dc2ccea50333a175b810#diff-878e6542838bfdd7b938be9cca336887R247).
